File: 2020/Python/4.py
passports = []
passport = {}
with open("4.txt") as inp:
    for line in inp:
        if line != "\n":
            line = line.rstrip().split(" ")
            line = [field.split(":") for field in line]
            for field in line:
                passport[field[0]] = field[1]
        else:
            passports.append(passport)
            passport = {}
    passports.append(passport)

validPassports = 0
for passport in passports:
    fieldCount = 0
    for key in passport.keys():
        fieldCount += 1
    if fieldCount == 8 or (fieldCount == 7 and (not "cid" in passport)):
        validPassports += 1
print(validPassports)

colors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth",]

def checkHeight(s):
    if s[-2:] == "cm":
        if 150 <= int(s[:-2]) <= 193:
            return True
        else:
            return False
    elif s[-2:] == "in":
        if 59 <= int(s[:-2]) <= 76:
            return True
        else:
            return False
    else:
        return False

validPassports = 0
for passport in passports:
    fieldCount = 0
    for key in passport.keys():
        fieldCount += 1
    if fieldCount == 8 or (fieldCount == 7 and (not "cid" in passport)):
        if (1920 <= int(passport["byr"]) <= 2002) \
        and (2010 <= int(passport["iyr"]) <= 2020) \
        and (2020 <= int(passport["eyr"]) <= 2030) \
        and (checkHeight(passport["hgt"]))\
        and ((passport["hcl"][0] == "#") and (len(passport["hcl"]) == 7)) \
        and (passport["ecl"] in colors) \
        and (passport["pid"].isnumeric() and len(passport["pid"]) == 9):
            validPassports += 1
print(validPassports)

File: 2020/Lua/4.lua
local pps = {}
local pp = {}
for line in io.lines("4.txt") do
  if line == "" then table.insert(pps, pp) pp = {} end
  for info in line:gmatch("%S+:%S+") do
    local k, v = info:match("(%S+)%:(%S+)")
    pp[k] = v
  end
end
table.insert(pps, pp)

local function checkByr(n)
  return tonumber(n) >= 1920 and tonumber(n) <= 2002 end

local function checkIyr(n)
  return tonumber(n) >= 2010 and tonumber(n) <= 2020 end

local function checkEyr(n)
  return tonumber(n) >= 2020 and tonumber(n) <= 2030 end

local function checkHgt(s)
  if s:sub(-2, -1) == "cm" then
    return tonumber(s:sub(1,-3)) >= 150 and tonumber(s:sub(1,-3)) <= 193
  elseif s:sub(-2, -1) == "in" then
    return tonumber(s:sub(1,-3)) >= 59 and tonumber(s:sub(1,-3)) <= 76
  end
end

local function checkHcl(s)
  return s:match("%#[0-9, a-f+]") and #s == 7 end

local function checkEcl(s)
  return s == "amb" or s == "blu" or s == "brn" or s == "gry" or s == "grn" or s == "hzl" or s == "oth" end

local function checkPid(s)
  return s:match("%d%d%d%d%d%d%d%d%d") and #s == 9 end

local count = 0
local count2 = 0
for _, pp in pairs(pps) do
  local c = 0
  for _, _ in pairs(pp) do
    c = c + 1
  end
  if (c == 8) or (c == 7 and not pp["cid"]) then
    count = count + 1
    if checkByr(pp["byr"]) and
      checkIyr(pp["iyr"]) and
      checkEyr(pp["eyr"]) and
      checkHgt(pp["hgt"]) and
      checkHcl(pp["hcl"]) and
      checkEcl(pp["ecl"]) and
      checkPid(pp["pid"]) then
        count2 = count2 + 1
      end
  end
end
print(count)
print(count2)
